src/glib/base/blobbs.cpp

```diff
@@ -508,47 +519,84 @@ TMBlobBs::~TMBlobBs(){
   }
 }
 
+// save a new buffer in SIn to a blob
 TBlobPt TMBlobBs::PutBlob(const PSIn& SIn){
   EAssert((Access==faCreate)||(Access==faUpdate)||(Access==faRestore));
-  TBlobPt BlobPt=SegV[CurSegN]->PutBlob(SIn);
-  if (BlobPt.Empty()){
-    for (uint16 SegN=0; SegN<SegV.Len(); SegN++){
-      BlobPt=SegV[CurSegN=SegN]->PutBlob(SIn);
-      if (!BlobPt.Empty()){break;}
+  // buffer size that we need to store
+  int BfL = SIn->Len();
+  // segment index from which we will start to check if we can insert data into
+  uint16 DestSegN = 0;
+  int BlockSize = 0;
+  for (int KeyId = BlockSizeToSegH.FFirstKeyId(); BlockSizeToSegH.FNextKeyId(KeyId);) {
+    BlockSize = BlockSizeToSegH.GetKey(KeyId);
+    // if we found a block size that could store the buffer
+    if (BlockSize > BfL) {
+      DestSegN = BlockSizeToSegH[KeyId];
+      break;
     }
-    if (BlobPt.Empty()){
-      TStr SegFNm=GetSegFNm(NrFPath, NrFMid, SegV.Len());
-      PBlobBs Seg=TGBlobBs::New(SegFNm, faCreate, MxSegLen);
-      CurSegN=SegV.Add(Seg); 
-      EAssert(CurSegN <= TUSInt::Mx);
-      BlobPt=SegV[CurSegN]->PutBlob(SIn);
+  }
+  // DestSegN should now be an index of segment from which we will start checking if the segment has place for the given data
+  // if not, the index will be increased
+  TBlobPt BlobPt;
+  while (DestSegN < SegV.Len()) {
+    BlobPt = SegV[DestSegN]->PutBlob(SIn);
+    if (!BlobPt.Empty()) { 
+      break; 
     }
+    DestSegN++;
+  }
+  // we were not able to find a segment of required size in any of the available segments
+  // create a new segment
+  if (BlobPt.Empty()) {
+    TStr SegFNm = GetSegFNm(NrFPath, NrFMid, SegV.Len());
+    PBlobBs Seg = TGBlobBs::New(SegFNm, faCreate, MxSegLen);
+    DestSegN = SegV.Add(Seg);
+    EAssert(DestSegN <= TUSInt::Mx);
```

Comment:
this warning doesn't make sense. uint16 is the data type for DestSegN. TUSInt has internal type uint